build: update build.ps1, use update-checkout for more dependencies

compnerd opened the pull request:
Update the build.ps1 to match the current state which introduces `-BuildTo` and updates most of the dependencies. Take the opportunity to migrate to update-checkout for the majority of the dependencies. The SQLite amalgamation and installer image are still fetched later, and ICU is still cloned in the CI wrapper.

compnerd committed:
build: update build.ps1, use update-checkout for more dependencies

Update the build.ps1 to match the current state which introduces
`-BuildTo` and updates most of the dependencies. Take the opportunity to
migrate to update-checkout for the majority of the dependencies. The
SQLite amalgamation and installer image are still fetched later, and ICU
is still cloned in the CI wrapper.

compnerd commented:
@swift-ci please build toolchain Windows platform

compnerd commented:
@swift-ci please build toolchain Windows platform

shahmishal commented on utils/build-windows-toolchain.bat:
```diff
@@ -123,13 +123,8 @@ setlocal enableextensions enabledelayedexpansion
-git clone --quiet --no-tags --depth 1 --branch v1.2.11 https://github.com/madler/zlib
-git clone --quiet --no-tags --depth 1 --branch v2.9.12 https://github.com/gnome/libxml2
-git clone --quiet --no-tags --depth 1 --branch version-3.36.0 https://github.com/sqlite/sqlite
+rd /s /q icu
 git clone --quiet --no-tags --depth 1 --branch maint/maint-69 https://github.com/unicode-org/icu
```
This does not look right, we are changing the version of the ICU.

compnerd replied:
This is not changing the version of ICU, ICU 69.1 is what Windows has used since 5.3.

shahmishal replied:
Another point, It's deleting local copy of the icu. This would be bad if an engineer had a ICU changes locally, they will lose the work.

shahmishal replied:
Now with @parkera @iCharlesHu feedback, we can drop the clone of icu from here.

shahmishal commented:
@parkera @iCharlesHu Are you ok with updating ICU tag to `maint/maint-69` as its required by Windows?

parkera commented:
@compnerd is there a way to verify that swift-corelibs-foundation's unit tests still pass with the updated ICU? Sometimes updates to formatting styles and such can cause issues.

iCharlesHu commented:
> @parkera @iCharlesHu Are you ok with updating ICU tag to `maint/maint-69` as its required for Windows?

If Windows has always used 69 I don't see an issue. Although I'd say we should definitely test swift-corelibs-foundation to make sure the new ICU doesn't break things.

compnerd commented:
@iCharlesHu, yeah, we have always used 69.3 for Windows. @parkera, I can stack the change to get ICU switched over to 69.1 and do a cross-repository test for Linux. I know that the build on Windows will also test Foundation as well.

compnerd committed:
build: update update-checkout to ICU 69

Unify the non-Darwin targets to ICU 69. Windows has been using ICU 69
since 5.3, Linux was at 65.

shahmishal commented on utils/update_checkout/update-checkout-config.json:
```diff
@@ -134,7 +141,11 @@
                 "swift-nio-ssl": "2.15.0",
                 "swift-experimental-string-processing": "swift/main",
-                "wasi-libc": "wasi-sdk-20"
+                "wasi-libc": "wasi-sdk-20",
+                "curl": "curl-8_4_0",
+                "icu": "maint/maint-69",
+                "libxml2": "v2.11.5",
+                "zlib": "v1.3"
```
Please also update the `next` scheme

compnerd committed:
update_checkout: update the next scheme to match main

Update next to match main for dependencies. This includes:
- CURL
- ICU
- libxml2
- zlib

compnerd commented:
@swift-ci please test

compnerd commented:
@parkera https://ci.swift.org/job/swift-corelibs-foundation-PR-Linux/306/ is the test run - seems that it doesn't impact the tests.

shahmishal commented:
We will need to have these changes cherry-picked into release/5.10.

compnerd commented:
@shahmishal I think that we want to get the changes into main and then cherry-pick. I can prepare a cherry-pick before it is merged of course if you prefer.

shahmishal commented:
> @shahmishal I think that we want to get the changes into main and then cherry-pick. I can prepare a cherry-pick before it is merged of course if you prefer.

@compnerd It can be done after, just wanted to make sure we cherry-pick these changes to support the nightly toolchain for release/5.10.

AnthonyLatsis commented on utils/update_checkout/update-checkout-config.json:
```diff
@@ -88,7 +88,16 @@
+        },
+        "zlib": {
+          "remote": { "id": "madler/zlib" }
+        }
```
Why do we suddenly need these, and icu, on macOS?

vanvoorden commented:
@compnerd Could this be related to https://github.com/apple/swift/issues/70966?

compnerd commented:
No, this only adds the repositories to the set, it doesn't write files into it.
